### PR TITLE
feat: adding aws metric streams input plugin

### PR DESCRIPTION
**Pull request by mccabecillian**
### Required for all PRs:

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

This PR creates a plugin to receive metrics from [AWS Metric Streams](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Metric-Streams.html).

The standard HTTP listener is incapable of meeting AWS requirements (decoding Base64 data, return request IDs, etc.) therefore a specific plugin was created for internal use at Groupon to listen for Firehose Delivery Stream data and parse the metrics. We are sharing the plugin that we created for the open source community to be able to leverage Metric Streams by AWS, and not rely on other services such as DataDog.

Full details of the plugin are available in the README and test data is available in the tests.

### CODE REVIEW

**Comment by mccabecillian** [comment]
Not entirely sure why the `lint-pr-changes` check is failing, if someone could check that for me to make the necessary changes

**Review by Hipska — CHANGES_REQUESTED** [review]

File: plugins/inputs/metric_streams_listener/metric_streams_listener.go
Comment (Hipska) [inline plugins/inputs/metric_streams_listener/metric_streams_listener.go:138]: Would it be possible to eliminate `tlsConf` from the struct?
```suggestion
	server.TLSConfig, err = ms.ServerConfig.TLSConfig()
	if err != nil {
		return err
	}

	if server.TLSConfig != nil {
		ms.listener, err = tls.Listen("tcp", ms.ServiceAddress, server.TLSConfig)
	} else {
		ms.listener, err = net.Listen("tcp", ms.ServiceAddress)
	}
	if err != nil {
		return err
	}
```

Comment (mccabecillian) [reply plugins/inputs/metric_streams_listener/metric_streams_listener.go]: Yeah nice, I am able to remove this - this change could also be added to the `http_listener_v2` plugin as well then, it seems like it is unnecessary to use as it is first only used to create the `http.Server` but then updated anyway based on the value returned by `ms.ServerConfig.TLSConfig()`.

**Comment by mccabecillian** [comment]
The README linter continues to fail due to other READMEs by the way, unrelated to the `metric_streams_listener` plugin.

**Comment by Hipska** [comment]
> The README linter continues to fail due to other READMEs by the way, unrelated to the `metric_streams_listener` plugin.

Yeah, never mind of that, its a known issue and we will ignore errors from unrelated files..

File: plugins/inputs/metric_streams_listener/metric_streams_listener.go
Comment (Hipska) [reply plugins/inputs/metric_streams_listener/metric_streams_listener.go]: Its known fact that some of the plugins might need a refactoring. This is something that will be part of Telegraf 2.0

**Comment by mccabecillian** [comment]
Hey @Hipska is there anything else required for this? I'm not sure what the approval & merge process is like, just wanted to check if I need to do anything else.

**Review by Hipska — APPROVED** [review]

File: plugins/inputs/metric_streams_listener/README.md
Comment (reimda) [inline plugins/inputs/metric_streams_listener/README.md:14]: This looks like valuable developer documentation, but maybe something users don't need to worry about. What do you think about moving them into the code as comments in the section where you decode the blocks and write the response?

File: plugins/inputs/metric_streams_listener/metric_streams_listener.go
Comment (reimda) [inline plugins/inputs/metric_streams_listener/metric_streams_listener.go:35]: I know you've already renamed this once, but I'm wondering if this is the best name for the plugin. Metric streams is part of cloudwatch, and we have a cloudwatch input plugin already. I'm not an AWS expert, so I'm not sure how much overlap there is between this new plugin and telegraf's existing cloudwatch plugin. Would it make sense to call this "cloudwatch metric streams" or "cloudwatch listener"?

Does the cloudwatch input handle the same aws metrics as this new plugin? My impression from quickly scanning the aws docs is that metric streams is a push model and cloudwatch is pull but they do handle the same source metrics.

If that's the case, I think this should have a similar name to the existing cloudwatch plugin so people get the impression that they do similar jobs.

File: plugins/inputs/metric_streams_listener/README.md
Comment (reimda) [inline plugins/inputs/metric_streams_listener/README.md:89]: Are the metrics produced by this plugin and the cloudwatch plugin the same? It would be a big usability win if they were because could switch from the cloudwatch plugin to this one without having to handle data model changes.

File: plugins/inputs/metric_streams_listener/metric_streams_listener.go
Comment (mccabecillian) [reply plugins/inputs/metric_streams_listener/metric_streams_listener.go:35]: I'm open to renaming it sure, is "cloudwatch_metric_streams_listener" too long?

File: plugins/inputs/metric_streams_listener/README.md
Comment (mccabecillian) [reply plugins/inputs/metric_streams_listener/README.md:14]: True, I can include it in the code with the response logic.

File: plugins/inputs/metric_streams_listener/README.md
Comment (mccabecillian) [reply plugins/inputs/metric_streams_listener/README.md:89]: We actually discussed this internally a bit, and we agree - we were going to make breaking changes in our own dashboards and alerts by conforming to the new naming after the migration from the CloudWatch API (pull method).

So I think you're right, and we should make this easier for anyone else who upgrades, so that we get that usability win you mentioned. So I will push an update to make the two align.

Comment (reimda) [reply plugins/inputs/metric_streams_listener/README.md:89]: That sounds great. I was in the middle of writing down my thoughts about it when you posted. Here it is anyway:

I think tweaking the names to match cloudwatch is the way to go if there are only a few differences.

If we don't make them compatible now, I expect that in the future someone will submit a PR to make them compatible. If we add compatibility at a future time, telegraf will need to make it optional so we don't break people who got used to the differences. Then the plugin will be stuck with maintaining two metric formats or face breaking compatibility for a portion of users.

If we do make them compatible now, we don't have to deal with two metric formats or breaking compatibility.

File: plugins/inputs/metric_streams_listener/metric_streams_listener.go
Comment (reimda) [reply plugins/inputs/metric_streams_listener/metric_streams_listener.go:35]: "cloudwatch_metric_streams_listener" is a mouthful. AWS's product branding word salad doesn't make it easy.

Many similar telegraf input plugins do have suffixes like "_listener" or "_consumer". There's even one case of "_push" with GCP's pubsub. The suffixes are used to disambiguate. For example "http" is pull, "http_listener" is push. "kafka" gathers metrics about the kafka server's performance and "kafka_consumer" receives kafka messages and parses them as metric data.

My vote is "cloudwatch_metric_streams" with no "_listener". It has a few things going for it:
* It's AWS's official product name
* It's similar to the existing "cloudwatch" plugin and will show up next to it in an alphabetical list.
* It's already not ambiguous. To use it, you have to be familiar with the AWS service. If you're familiar with the service you'll know "cloudwatch" is pull and "cloudwatch_metric_streams" is push.

**Comment by mccabecillian** [comment]
I made the changes as discussed @reimda, the plugin is renamed to `cloudwatch_metric_streams` and the fields are updated.

The necessary changes are made to the README too.

File: plugins/inputs/cloudwatch_metric_streams/README.md
Comment (mccabecillian) [reply plugins/inputs/cloudwatch_metric_streams/README.md:126]: Good point, I was just following an example from another README for posting data to the listener - I never realised I had no data actually there though.

I think I should just remove it?

Comment (Hipska) [reply plugins/inputs/cloudwatch_metric_streams/README.md:126]: Fine by me

File: plugins/inputs/cloudwatch_metric_streams/README.md
Comment (mccabecillian) [reply plugins/inputs/cloudwatch_metric_streams/README.md:64]: I personally think so because the actual data model was not readily available by AWS, I had to send data and the decode it.

So this JSON is just the metric itself, skipping the other details that outlined by AWS in their documentation.

Comment (Hipska) [reply plugins/inputs/cloudwatch_metric_streams/README.md:64]: Okay, but what about the references to  `data`, `record` or `records` then?

File: plugins/inputs/cloudwatch_metric_streams/README.md
Comment (mccabecillian) [reply plugins/inputs/cloudwatch_metric_streams/README.md:79]: A point made by @reimda was to just make the two compatible so that we do not have to handle two different formats.

Do you think we should handle two formats?

File: plugins/inputs/cloudwatch_metric_streams/README.md
Comment (mccabecillian) [reply plugins/inputs/cloudwatch_metric_streams/README.md:64]: They are outlined here: https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html

Do you think I should add this here?

There is also a full sample message in the tests, but it's a bit too long to add to the README.

File: plugins/inputs/metric_streams_listener/README.md
Comment (Hipska) [reply plugins/inputs/metric_streams_listener/README.md]: Oh, missed this.

@reimda what about users not used to the other plugin, wouldn't they find it weird the names from metric stream get renamed here? IMHO it could be a config flag to choose..

File: plugins/inputs/cloudwatch_metric_streams/README.md
Comment (mccabecillian) [reply plugins/inputs/cloudwatch_metric_streams/README.md:79]: No no, all good to me, we can wrap this is a conditional for sure

File: plugins/inputs/cloudwatch_metric_streams/README.md
Comment (Hipska) [reply plugins/inputs/cloudwatch_metric_streams/README.md:64]: I would mention this sample JSON is what comes from the decoded `data` field. Or didn't I understand the docs correctly?

File: plugins/inputs/cloudwatch_metric_streams/README.md
Comment (mccabecillian) [reply plugins/inputs/cloudwatch_metric_streams/README.md:79]: I added the flag anyway, we can always revert if there is a good reason not to but choice is probably better.

File: plugins/inputs/cloudwatch_metric_streams/README.md
Comment (mccabecillian) [reply plugins/inputs/cloudwatch_metric_streams/README.md:64]: No that's right, so a request from AWS looks like:

```json
{
  "requestId": "ed4acda5-034f-9f42-bba1-f29aea6d7d8f",
  "timestamp": 1578090901599,
  "records": [
   {
    "data": {
```

And the `data` is one Base64 encoded string, that is actually multiple JSON blocks.

I think it might be a bit too much developer information to add all of that, but I will update the README to include that the JSON is first Base64 encoded within `data`.

**Comment by mccabecillian** [comment]
Yep.. I forgot to run it locally before pushing:

```bash
make lint-install
make lint-branch
```

File: plugins/inputs/cloudwatch_metric_streams/README.md
Comment (Hipska) [inline plugins/inputs/cloudwatch_metric_streams/README.md:103]: Could  you also add an example of the default mode

File: plugins/inputs/cloudwatch_metric_streams/README.md
Comment (Hipska) [inline plugins/inputs/cloudwatch_metric_streams/README.md:91]: When compatibility is enabled.

**Comment by mccabecillian** [comment]
Hey @Hipska & @reimda any updates on this? I think we're good to merge, we have deployed this version of the plugin in our own environment without issues.

**Comment by Hipska** [comment]
!retry-checks

**Comment by Hipska** [comment]
!retry-failed

File: plugins/inputs/metric_streams_listener/README.md
Comment (reimda) [reply plugins/inputs/metric_streams_listener/README.md]: You have a good point. Let's use the same format as the other plugin for this PR and if users request support for the native metric streams naming we can look into adding a setting in another PR.

(I hope we don't end up making both plugins compatible with both naming variants in the future.)

**Comment by reimda** [comment]
Hi @mccabecillian the code in the PR looks good to me. There are still errors in CI that we need to clear up before merging. Could you rebase this branch on master and force push? I think the errors will go away if you rebase. Thanks!

**Comment by mccabecillian** [comment]
CI is all green now 🎉 

Let me know if anything else is needed @reimda & @Hipska

**Comment by reimda** [comment]
Thanks @mccabecillian! Github says this is your first contribution to telegraf. Great job! We would love to work with you again sometime.

**Pull request merged**
